src/core/lib/iomgr/executor/mpmcqueue.cc

```diff
@@ -0,0 +1,114 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/iomgr/executor/mpmcqueue.h"
+
+namespace grpc_core {
+
+DebugOnlyTraceFlag grpc_thread_pool_trace(false, "thread_pool");
+
+inline void* InfLenFIFOQueue::PopFront() {
+  // Caller should already check queue is not empty and has already held the
+  // mutex. This function will only do the job of removal.
+  void* result = queue_head_->content;
+  Node* head_to_remove = queue_head_;
+  queue_head_ = queue_head_->next;
+
+  count_.FetchSub(1, MemoryOrder::RELAXED);
```

Comment:
As mentioned offline, assuming count_ is not modified outside of lock this wants to be load, subtract 1, store. FetchAdd/FetchSub are more expensive (even relaxed) than load-add-store because they guarantee atomicity.